File: apps/api/src/features/enrollment/utils/canonical-message.ts
import { CanonicalMessage } from "../entities/generators";

/**
 * Fixed field order for the pipe-separated canonical message signed by the
 * device's Keystore key at enrollment. Must exactly match the Android
 * agent's serialization — do NOT derive this from Object.values/key
 * insertion order, since that's not guaranteed to stay aligned between the
 * Kotlin and TypeScript sides as fields get added.
 *
 * Used on both sides: the agent (and scripts/mock-device-enroll.ts) build
 * this string to sign, and the server rebuilds the exact same string to
 * verify the signature (see device/service.ts).
 */
const CANONICAL_FIELD_ORDER: (keyof CanonicalMessage)[] = [
    "model",
    "manufacturer",
    "release",
    "serialNumber",
    "imei",
    "macAddress",
    "androidId",
    "method",
    "timestamp",
    "publicKey",
    "challenge",
];

export const generateCanonicalMessage = (request: CanonicalMessage): string =>
    CANONICAL_FIELD_ORDER.map((field) => request[field] ?? "").join("|");

File: apps/android-agent/app/src/main/java/com/openmdm/agent/security/CanonicalMessage.kt
package com.openmdm.agent.security

/**
 * Builds the pipe-separated canonical message signed (proof of possession) at
 * enrollment. Field order and semantics MUST exactly match the server's
 * `generateCanonicalMessage` (apps/api/src/features/enrollment/utils/canonical-message.ts):
 *
 *   model|manufacturer|release|serialNumber|imei|macAddress|androidId|method|timestamp|publicKey|challenge
 *
 * An absent optional field is an empty string ("") in the message, never
 * omitted and never the literal "null" — see [CanonicalMessage.build].
 *
 * Kept free of any Android/AndroidKeyStore dependency so it's testable with
 * plain JUnit on the host JVM.
 */
object CanonicalMessage {

    fun build(
        model: String,
        manufacturer: String,
        release: String,
        serialNumber: String?,
        imei: String?,
        macAddress: String?,
        androidId: String?,
        method: String?,
        timestamp: String,
        publicKey: String,
        challenge: String,
    ): String = listOf(
        model,
        manufacturer,
        release,
        serialNumber.orEmpty(),
        imei.orEmpty(),
        macAddress.orEmpty(),
        androidId.orEmpty(),
        method.orEmpty(),
        timestamp,
        publicKey,
        challenge,
    ).joinToString("|")
}
